File: src/Routing/ShortestPath/BidirectDijkstra.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Simple.GeoData;
using Simple.Routing.Graph;

namespace Simple.Routing.ShortestPath
{
    class BidirectDijkstra : IShortestPath
    {
        private PriorityQueue<int, int> startheap;
        private PriorityQueue<int, int> endheap;
        private int midid;
        private int startid;
        private int endid;
        private IGraph graph;
        private IGeometry geom;
        private IWeighting weight;
        private Flag[] flags;

        private struct Flag
        {
            public double pathlength;
            public double pathlength2;
            public int prevEdge;
            public int prevEdge2;
            public bool visited;
            public bool visited2;
        }

        public BidirectDijkstra(IGraph graph, int start, int end)
        {
            this.graph = graph;
            this.startid = start;
            this.endid = end;
            this.startheap = new PriorityQueue<int, int>();
            this.startheap.Enqueue(this.startid, 0);
            this.endheap = new PriorityQueue<int, int>();
            this.endheap.Enqueue(this.endid, 0);
            this.flags = new Flag[graph.nodeCount()];
            this.geom = graph.getGeometry();
            this.weight = graph.getWeighting();
            for (int i = 0; i < flags.Length; i++)
            {
                flags[i].pathlength = 1000000000;
                flags[i].pathlength2 = 1000000000;
            }
            flags[start].pathlength = 0;
            flags[end].pathlength2 = 0;
        }

        private bool finished = false;
        /// <summary>
        /// performs one step of algorithm (one step from each direction)
        /// </summary>
        /// <returns>false if shortest path found</returns>
        public bool calcShortestPath()
        {
            var s = Task.Run(() => { fromStart(); });
            var e = Task.Run(() => { fromEnd(); });
            Task.WaitAll(new[] {s,e});
            return true;
        }

        private void fromStart()
        {
            int currid;
            while (!finished)
            {
                currid = this.startheap.Dequeue();
                ref NodeAttributes curr = ref this.graph.getNode(currid);
                ref Flag currflag = ref this.flags[currid];
                if (currflag.visited)
                {
                    continue;
                }
                if (currflag.visited2)
                {
                    midid = currid;
                    finished = true;
                    return;
                }
                currflag.visited = true;
                IEdgeRefStore edges = this.graph.getAdjacentEdges(currid);
                for (int i = 0; i < edges.length; i++)
                {
                    int edgeid = edges[i];
                    ref EdgeAttributes edge = ref this.graph.getEdge(edgeid);
                    int otherid = this.graph.getOtherNode(edgeid, currid, out Direction dir);
                    ref NodeAttributes other = ref this.graph.getNode(otherid);
                    ref Flag otherflag = ref this.flags[otherid];
                    if (otherflag.visited)
                    {
                        continue;
                    }
                    if (edge.oneway && dir == Direction.backward)
                    {
                        continue;
                    }
                    double newlength = currflag.pathlength + this.weight.getEdgeWeight(edgeid);
                    if (otherflag.pathlength > newlength)
                    {
                        otherflag.prevEdge = edgeid;
                        otherflag.pathlength = newlength;
                        this.startheap.Enqueue(otherid, (int)newlength);
                    }
                }
            }
        }

        private void fromEnd()
        {
            int currid;
            while (!finished)
            {
                currid = this.endheap.Dequeue();
                ref NodeAttributes curr = ref this.graph.getNode(currid);
                ref Flag currflag = ref this.flags[currid];
                if (currflag.visited2)
                {
                    continue;
                }
                if (currflag.visited)
                {
                    midid = currid;
                    finished = true;
                    return;
                }
                currflag.visited2 = true;
                IEdgeRefStore edges = this.graph.getAdjacentEdges(currid);
                for (int i = 0; i < edges.length; i++)
                {
                    int edgeid = edges[i];
                    ref EdgeAttributes edge = ref this.graph.getEdge(edgeid);
                    int otherid = this.graph.getOtherNode(edgeid, currid, out Direction dir);
                    ref NodeAttributes other = ref this.graph.getNode(otherid);
                    ref Flag otherflag = ref this.flags[otherid];
                    if (otherflag.visited2)
                    {
                        continue;
                    }
                    if (edge.oneway && dir == Direction.forward)
                    {
                        continue;
                    }
                    double newlength = currflag.pathlength2 + this.weight.getEdgeWeight(edgeid);
                    if (otherflag.pathlength2 > newlength)
                    {
                        otherflag.prevEdge2 = edgeid;
                        otherflag.pathlength2 = newlength;
                        this.endheap.Enqueue(otherid, (int)newlength);
                    }
                }
            }
        }

        private int currid;
        public bool steps(int count, List<ICoordArray> visitededges)
        {
            for (int c = 0; c < count; c++)
            {
                currid = this.startheap.Dequeue();
                ref NodeAttributes curr = ref this.graph.getNode(currid);
                ref Flag currflag = ref this.flags[currid];
                if (currflag.visited)
                {
                    continue;
                }
                if (currflag.visited2)
                {
                    midid = currid;
                    return false;
                }
                currflag.visited = true;
                IEdgeRefStore edges = this.graph.getAdjacentEdges(currid);
                for (int i = 0; i < edges.length; i++)
                {
                    int edgeid = edges[i];
                    ref EdgeAttributes edge = ref this.graph.getEdge(edgeid);
                    int otherid = this.graph.getOtherNode(edgeid, currid, out Direction dir);
                    ref NodeAttributes other = ref this.graph.getNode(otherid);
                    ref Flag otherflag = ref this.flags[otherid];
                    if (otherflag.visited)
                    {
                        continue;
                    }
                    if (edge.oneway && dir == Direction.backward)
                    {
                        continue;
                    }
                    visitededges.Add(this.geom.getEdge(edgeid));
                    double newlength = currflag.pathlength + this.weight.getEdgeWeight(edgeid);
                    if (otherflag.pathlength > newlength)
                    {
                        otherflag.prevEdge = edgeid;
                        otherflag.pathlength = newlength;
                        this.startheap.Enqueue(otherid, (int)newlength);
                    }
                }
                currid = this.endheap.Dequeue();
                curr = ref this.graph.getNode(currid);
                currflag = ref this.flags[currid];
                if (currflag.visited2)
                {
                    continue;
                }
                if (currflag.visited)
                {
                    midid = currid;
                    return false;
                }
                currflag.visited2 = true;
                edges = this.graph.getAdjacentEdges(currid);
                for (int i = 0; i < edges.length; i++)
                {
                    int edgeid = edges[i];
                    ref EdgeAttributes edge = ref this.graph.getEdge(edgeid);
                    int otherid = this.graph.getOtherNode(edgeid, currid, out Direction dir);
                    ref NodeAttributes other = ref this.graph.getNode(otherid);
                    ref Flag otherflag = ref this.flags[otherid];
                    if (otherflag.visited2)
                    {
                        continue;
                    }
                    if (edge.oneway && dir == Direction.forward)
                    {
                        continue;
                    }
                    visitededges.Add(this.geom.getEdge(edgeid));
                    double newlength = currflag.pathlength2 + this.weight.getEdgeWeight(edgeid);
                    if (otherflag.pathlength2 > newlength)
                    {
                        otherflag.prevEdge2 = edgeid;
                        otherflag.pathlength2 = newlength;
                        this.endheap.Enqueue(otherid, (int)newlength);
                    }
                }
            }
            return true;
        }

        /// <summary>
        /// use only after path finsing finished
        /// </summary>
        /// <returns>list of LineD representing shortest path</returns>
        public Path getShortestPath()
        {
            List<int> path = new List<int>();
            int edge;
            currid = midid;
            while (true)
            {
                path.Add(currid);
                if (currid == startid)
                {
                    break;
                }
                edge = this.flags[currid].prevEdge;
                path.Add(edge);
                currid = this.graph.getOtherNode(edge, currid, out Direction _);
            }
            path.RemoveAt(0);
            path.Reverse();
            currid = midid;
            while (true)
            {
                path.Add(currid);
                if (currid == endid)
                {
                    break;
                }
                edge = this.flags[currid].prevEdge2;
                path.Add(edge);
                currid = this.graph.getOtherNode(edge, currid, out Direction _);
            }
            return new Path(this.graph, path);
        }
    }
}

File: src/go-routing/routing/bidirect_dijkstra.go
package routing

import (
	"fmt"

	"github.com/ttpr0/simple-routing-visualizer/src/go-routing/geo"
	"github.com/ttpr0/simple-routing-visualizer/src/go-routing/graph"
	. "github.com/ttpr0/simple-routing-visualizer/src/go-routing/util"
)

type flag_bd struct {
	path_length1 float64
	path_length2 float64
	prev_edge1   int32
	prev_edge2   int32
	visited1     bool
	visited2     bool
}

type BidirectDijkstra struct {
	startheap PriorityQueue[int32, float64]
	endheap   PriorityQueue[int32, float64]
	mid_id    int32
	start_id  int32
	end_id    int32
	graph     graph.IGraph
	geom      graph.IGeometry
	weight    graph.IWeighting
	flags     []flag_bd
}

func NewBidirectDijkstra(graph graph.IGraph, start, end int32) *BidirectDijkstra {
	d := BidirectDijkstra{graph: graph, start_id: start, end_id: end, geom: graph.GetGeometry(), weight: graph.GetWeighting()}

	flags := make([]flag_bd, graph.NodeCount())
	for i := 0; i < len(flags); i++ {
		flags[i].path_length1 = 1000000000
		flags[i].path_length2 = 1000000000
	}
	flags[start].path_length1 = 0
	flags[end].path_length2 = 0
	d.flags = flags

	startheap := NewPriorityQueue[int32, float64](100)
	startheap.Enqueue(d.start_id, 0)
	d.startheap = startheap
	endheap := NewPriorityQueue[int32, float64](100)
	endheap.Enqueue(d.end_id, 0)
	d.endheap = endheap

	return &d
}

func (self *BidirectDijkstra) CalcShortestPath() bool {
	finished := false
	for !finished {
		// from start
		curr_id, _ := self.startheap.Dequeue()
		//curr := (*d.graph).GetNode(curr_id)
		curr_flag := self.flags[curr_id]

		if curr_flag.visited1 {
			continue
		}
		curr_flag.visited1 = true
		edges := self.graph.GetAdjacentEdges(curr_id, graph.FORWARD)
		for {
			ref, ok := edges.Next()
			if !ok {
				break
			}
			if !ref.IsEdge() {
				continue
			}
			edge_id := ref.EdgeID
			other_id := ref.OtherID
			//other := (*d.graph).GetNode(other_id)
			other_flag := self.flags[other_id]

			if other_flag.visited1 {
				continue
			}

			new_length := curr_flag.path_length1 + float64(self.weight.GetEdgeWeight(edge_id))

			if other_flag.visited2 {
				shortest := new_length + other_flag.path_length2
				top1, ok1 := self.startheap.Peek()
				top2, ok2 := self.endheap.Peek()
				if ok1 && ok2 && float64(top1+top2) >= shortest {
					other_flag.prev_edge1 = edge_id
					other_flag.path_length1 = new_length
					self.flags[other_id] = other_flag
					self.mid_id = other_id
					finished = true
					break
				}
			}

			if other_flag.path_length1 > new_length {
				other_flag.prev_edge1 = edge_id
				other_flag.path_length1 = new_length
				self.startheap.Enqueue(other_id, new_length)
			}
			self.flags[other_id] = other_flag
		}
		self.flags[curr_id] = curr_flag

		if finished {
			break
		}
		// from end
		curr_id, _ = self.endheap.Dequeue()
		//curr := (*d.graph).GetNode(curr_id)
		curr_flag = self.flags[curr_id]

		if curr_flag.visited2 {
			continue
		}
		curr_flag.visited2 = true
		edges = self.graph.GetAdjacentEdges(curr_id, graph.BACKWARD)
		for {
			ref, ok := edges.Next()
			if !ok {
				break
			}
			if !ref.IsEdge() {
				continue
			}
			edge_id := ref.EdgeID
			other_id := ref.OtherID
			//other := (*d.graph).GetNode(other_id)
			other_flag := self.flags[other_id]

			if other_flag.visited2 {
				continue
			}

			new_length := curr_flag.path_length2 + float64(self.weight.GetEdgeWeight(edge_id))

			if other_flag.visited1 {
				shortest := new_length + other_flag.path_length1
				top1, ok1 := self.startheap.Peek()
				top2, ok2 := self.endheap.Peek()
				if ok1 && ok2 && float64(top1+top2) >= shortest {
					other_flag.prev_edge2 = edge_id
					other_flag.path_length2 = new_length
					self.flags[other_id] = other_flag
					self.mid_id = other_id
					finished = true
					break
				}
			}

			if other_flag.path_length2 > new_length {
				other_flag.prev_edge2 = edge_id
				other_flag.path_length2 = new_length
				self.endheap.Enqueue(other_id, new_length)
			}
			self.flags[other_id] = other_flag
		}
		self.flags[curr_id] = curr_flag
	}
	return true
}

func (self *BidirectDijkstra) Steps(count int, visitededges *List[geo.CoordArray]) bool {
	for c := 0; c < count; c++ {
		curr_id, _ := self.startheap.Dequeue()
		//curr := (*d.graph).GetNode(curr_id)
		curr_flag := self.flags[curr_id]
		if curr_flag.visited1 {
			continue
		}
		curr_flag.visited1 = true
		edges := self.graph.GetAdjacentEdges(curr_id, graph.FORWARD)
		for {
			ref, ok := edges.Next()
			if !ok {
				break
			}
			if !ref.IsEdge() {
				continue
			}
			edge_id := ref.EdgeID
			other_id := ref.OtherID
			//other := (*d.graph).GetNode(other_id)
			other_flag := self.flags[other_id]
			if other_flag.visited1 {
				continue
			}
			visitededges.Add(self.geom.GetEdge(edge_id))
			new_length := curr_flag.path_length1 + float64(self.weight.GetEdgeWeight(edge_id))
			if other_flag.visited2 {
				shortest := new_length + other_flag.path_length2
				top1, ok1 := self.startheap.Peek()
				top2, ok2 := self.endheap.Peek()
				if ok1 && ok2 && float64(top1+top2) >= shortest {
					other_flag.prev_edge1 = edge_id
					other_flag.path_length1 = new_length
					self.flags[other_id] = other_flag
					self.mid_id = other_id
					return false
				}
			}
			if other_flag.path_length1 > new_length {
				other_flag.prev_edge1 = edge_id
				other_flag.path_length1 = new_length
				self.startheap.Enqueue(other_id, new_length)
			}
			self.flags[other_id] = other_flag
		}
		self.flags[curr_id] = curr_flag

		curr_id, _ = self.endheap.Dequeue()
		//curr := (*d.graph).GetNode(curr_id)
		curr_flag = self.flags[curr_id]
		if curr_flag.visited2 {
			continue
		}
		curr_flag.visited2 = true
		edges = self.graph.GetAdjacentEdges(curr_id, graph.BACKWARD)
		for {
			ref, ok := edges.Next()
			if !ok {
				break
			}
			if !ref.IsEdge() {
				continue
			}
			edge_id := ref.EdgeID
			other_id := ref.OtherID
			//other := (*d.graph).GetNode(other_id)
			other_flag := self.flags[other_id]
			if other_flag.visited2 {
				continue
			}
			visitededges.Add(self.geom.GetEdge(edge_id))
			new_length := curr_flag.path_length2 + float64(self.weight.GetEdgeWeight(edge_id))
			if other_flag.visited1 {
				shortest := new_length + other_flag.path_length1
				top1, ok1 := self.startheap.Peek()
				top2, ok2 := self.endheap.Peek()
				if ok1 && ok2 && float64(top1+top2) >= shortest {
					other_flag.prev_edge2 = edge_id
					other_flag.path_length2 = new_length
					self.flags[other_id] = other_flag
					self.mid_id = other_id
					return false
				}
			}
			if other_flag.path_length2 > new_length {
				other_flag.prev_edge2 = edge_id
				other_flag.path_length2 = new_length
				self.endheap.Enqueue(other_id, new_length)
			}
			self.flags[other_id] = other_flag
		}
		self.flags[curr_id] = curr_flag
	}
	return true
}

func (self *BidirectDijkstra) GetShortestPath() Path {
	path := make([]int32, 0, 10)
	length := int32(self.flags[self.mid_id].path_length1 + self.flags[self.mid_id].path_length2)
	curr_id := self.mid_id
	var edge int32
	for {
		if curr_id == self.start_id {
			break
		}
		edge = self.flags[curr_id].prev_edge1
		path = append(path, edge)
		curr_id, _ = self.graph.GetOtherNode(edge, curr_id)
	}
	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
		path[i], path[j] = path[j], path[i]
	}
	curr_id = self.mid_id
	for {
		if curr_id == self.end_id {
			break
		}
		edge = self.flags[curr_id].prev_edge2
		path = append(path, edge)
		curr_id, _ = self.graph.GetOtherNode(edge, curr_id)
	}
	fmt.Println("length:", length)
	return NewPath(self.graph, path)
}
